go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -291,86 +470,184 @@ func (vc *vcopier) copyTable(ctx context.Context, tableName string, copyState ma
 				TableName: initialPlan.SendRule.Match,
 			}
 			fieldEvent.Fields = append(fieldEvent.Fields, rows.Fields...)
-			vc.tablePlan, err = plan.buildExecutionPlan(fieldEvent)
+			tablePlan, err := plan.buildExecutionPlan(fieldEvent)
 			if err != nil {
 				return err
 			}
 			pkfields = append(pkfields, rows.Pkfields...)
 			buf := sqlparser.NewTrackedBuffer(nil)
-			buf.Myprintf("insert into _vt.copy_state (lastpk, vrepl_id, table_name) values (%a, %s, %s)", ":lastpk", strconv.Itoa(int(vc.vr.id)), encodeString(tableName))
-			addLatestCopyState = buf.ParsedQuery()
+			buf.Myprintf(
+				"insert into _vt.copy_state (lastpk, vrepl_id, table_name) values (%a, %s, %s)", ":lastpk",
+				strconv.Itoa(int(vc.vr.id)),
+				encodeString(tableName))
+			addLatestCopyState := buf.ParsedQuery()
+			copyWorkQueue.open(addLatestCopyState, pkfields, tablePlan)
 		}
 		if len(rows.Rows) == 0 {
 			return nil
 		}
 
-		// The number of rows we receive depends on the packet size set
-		// for the row streamer. Since the packet size is roughly equivalent
-		// to data size, this should map to a uniform amount of pages affected
-		// per statement. A packet size of 30K will roughly translate to 8
-		// mysql pages of 4K each.
-		if err := vc.vr.dbClient.Begin(); err != nil {
-			return err
+		// Clone rows, since pointer values will change while async work is
+		// happening. Can skip this when there's no parallelism.
+		if parallelism > 1 {
+			rows = proto.Clone(rows).(*binlogdatapb.VStreamRowsResponse)
 		}
-		_, err = vc.tablePlan.applyBulkInsert(&sqlbuffer, rows, func(sql string) (*sqltypes.Result, error) {
-			start := time.Now()
 
-			qr, err := vc.vr.dbClient.ExecuteWithRetry(ctx, sql)
-			if err != nil {
-				return nil, err
-			}
-			vc.vr.stats.QueryTimings.Record("copy", start)
-			vc.vr.stats.CopyRowCount.Add(int64(qr.RowsAffected))
-			vc.vr.stats.QueryCount.Add("copy", 1)
-			return qr, err
-		})
-		if err != nil {
-			return err
+		// Prepare a vcopierCopyTask for the current batch of work.
+		// TODO(maxeng) see if using a pre-allocated pool will speed things up.
+		currCh := make(chan *vcopierCopyTaskResult, 1)
+		currT := newVCopierCopyTask(taskID, newVCopierCopyTaskArgs(rows.Rows, rows.Lastpk))
+		taskID++
```

Comment:
I'm not sure. Right now `taskID` isn't used for anything. I had been using it during development for debugging. Going to remove it in next commit to avoid any confusion.